### PR TITLE
feat: implement automatic `AssignedData` rotation

### LINKED ISSUE #256: Implement automatic `AssignedData` rotation
## Description

To ensure the `AssignedData` is always up-to-date, implement a feature that automatically checks and updates the assignment table's `endDate` when accessed. This feature should be applied to all CRUD endpoints related to `AssignedData`.

## Criteria

- [x] Generate new `AssignedData` when the `endDate` is already reached
- [x] Make sure it generates correctly rotated `AssignedData` even if there are long gaps

## Notes
- All CRUD endpoints that require `AssignedData` check the endDate upon access.
- If the `endDate` has passed, a new `AssignedData` is generated before any CRUD operation is performed.
- The system can handle and correctly generate a new `AssignedData` even if there are long gaps between accesses.

### PR DESCRIPTION
**Pull request by akiomatic**
## Overview

In this PR, I introduce automatic rotation logic for `AssignedData` to ensure that the data remains up-to-date. The logic checks whether the end date of the current `AssignedData` has passed and, if so, generates the next rotation's `AssignedData` with the current categories, tasks, and tenants data. This also manages long gaps between accesses and ensures that tenants receive the most current and fair assignments. Additionally, tenants who need to take on extra categories are fairly assigned using the `extraAssignedCount` of each tenant, and the count is updated accordingly.

## Changes

`src/utils/automatic-rotation.ts`
- Implemented logic to rotate `AssignedData` when its end date has passed, ensuring the gap between the end date and the current date is handled appropriately

`src/app/api/[[...route]]/middlewares/automatic-rotation.middleware.ts`
- Added a middleware that automatically rotates the `AssignedData` for each share house owned by the logged-in landlord, if necessary, using the logic defined in `automatic-rotation.ts`
- This middleware is invoked on the following routes:
  -  /api/sharehouse
  - /api/sharehouses
  -  /api/rotation
  -  /api/category
  -  /api/task
  -  /api/tenant
  -  /api/whoami

`src/app/api/[[...route]]/routes/assignments.route.ts`
- `GET /api/assignments/:assignmentSheetId/:tenantId`
  - Integrated the automatic rotation logic to rotate the `AssignedData` if needed before returning the tenant's tasks
- `PATCH /api/assignments/:assignmentSheetId/:tenantId`
  - Added a check to determine if the associated `AssignedData`'s end date has passed. If it has, an error message stating "The assignment sheet is outdated. Please refresh the page." is returned, and the request is rejected. If the end date has not passed, the request proceeds as usual.

## Review points

Please manually modify the end date in Prisma Studio and verify that the automatic rotation is functioning correctly in your environment, especially in scenarios with long gaps between accesses. Additionally, please review the Tenants table to ensure that the `extraAssignedCount` is being updated accurately.

### Landlord's POV

1. When the end date of the current rotation has passed:
https://www.youtube.com/watch?v=DWZB7d6TTJo

### Tenant's POV

1. When the end date of the current rotation has passed:
https://www.youtube.com/watch?v=4-XrgeDleOU

2. When attempting to update the task status of the old rotation:
https://www.youtube.com/watch?v=AhQZOiyWu6M

## Assignee Checklist:

- [x] The base branch is correct (no accidental merges)
- [x] The branch name follows our branch naming rules
- [x] The PR title follows our PR title rules
- [x] My code follows our coding style
- [x] I have commented on my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings or errors in the console

## Reviewer Checklist:

- [ ] Code readability and simplicity
- [ ] Follows best practices and coding standards
- [ ] Understandable and maintainable code

### CODE REVIEW

File: src/constants/server-error-messages.ts [hunk 1]
```diff
@@ -35,4 +35,6 @@ export const SERVER_ERROR_MESSAGES = {
   NOT_EXISTING_USER: 'Email does not exist.',
   EXISTING_EMAIL: 'Email already in use. please log in.',
   CREDENTIAL_FIELDS_INVALID: 'Invalid fields',
+  PAST_END_DATE_ERROR:
+    'The assignment sheet is outdated. Please refresh the page.',
```
Comment (KaiKoide) [inline src/constants/server-error-messages.ts]: To my understanding, the tenants don't understand what the assignment sheet is because there is no explanation provided for it. Could you rephrase the word for another expression?

**Comment by KaiKoide** [comment]
Thank you for your work! I have a question about `extraAssignedCount`. In the video, I changed the end date in Prisma Studio, and `extraAssignedCount` for Tenant One increased, but not for Tenant Two, which was assigned 2 categories in the next rotation. Is this expected behavior?
https://youtu.be/70tkS725Ino

**Comment by akiomatic** [comment]
@KaiKoide Thank you for your review. I'll answer your question.

> extraAssignedCount for Tenant One increased, but not for Tenant Two, which was assigned 2 categories in the next rotation. Is this expected behavior?

Initially, I also thought there might be an issue, however, <ins>I realized this behavior is indeed as expected.</ins>

There are two reasons why this confusion may have occurred:

1. Throughout the demo, you were seeing the "next" rotation's task assignments.
2. The seed file didn't update the `extraAssignedCount` when creating the 1st rotation's task assignments, even thought a tenant had extra tasks.

Here’s what happened in the demo video:

1. (**The 1st rotation**) Tenant 1 was assigned to 2 categories. Since their `extraAssignedCount` wasn't updated to 1 in the seed file, that's why you see that Tenant 1 has 2 categories assigned in both the 1st (current) and the 2nd (next) rotations.
2. You modified the end date and updated the `AssignedData` accordingly.
3. (**The 2nd rotation**) Tenant 1 again became the tenant who is assigned to 2 categories. However, in this case, the `extraAssignedCount` was correctly updated to 1.
4. You were looking at the "next" rotation's task assignments, meaning it was for the **3rd** rotation (though in theory, it was actually the 2nd rotation’s task assignments).

image

**Comment by akiomatic** [comment]
@KaiKoide I've updated the error message based on your suggestion. Could you re-review this PR? Thank you.

**Review by KaiKoide — APPROVED** [review]
LGTM! Thank you!
